sys/include/usb/usbus.h

```diff
@@ -0,0 +1,515 @@
+                                           registration)                     */
+    struct usbus_interface_alt *alts; /**< List of alt settings              */
+    usbus_handler_t* handler;         /**< Handlers for this interface       */
+    usbus_hdr_gen_t *hdr_gen;         /**< Optional additional header
+                                           generators                        */
+    usbus_endpoint_t *ep;             /**< Linked list of endpoints belonging
+                                           to this interface                 */
+    usbus_string_t *descr;            /**< Descriptor string                 */
+    uint16_t idx;                     /**< Interface index, (set by USBUS
+                                           during registration               */
+    uint8_t class;                    /**< USB interface class               */
+    uint8_t subclass;                 /**< USB interface subclass            */
+    uint8_t protocol;                 /**< USB interface protocol            */
+} usbus_interface_t;
+
+/**
+ * @brief USBUS event handler function pointers
+ */
+typedef struct usbus_handler_driver{
+
+    /**
+     * @brief Initialize the event handler
+     *
+     * This function is called in the USBUS thread context to initialize the
+     * event handler
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     */
+    void (*init)(usbus_t *usbus, struct usbus_handler *handler);
+
+    /**
+     * @brief event handler function
+     *
+     * This function is passed USBUS events
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     * @param event     @ref USBUS_MSG_TYPES "event" to handle
+     * @param arg       Additional argument
+     */
+    int (*event_handler)(usbus_t *usbus, struct usbus_handler *handler, uint16_t event, void *arg);
+} usbus_handler_driver_t;
+
+/**
+ * @ USBUS handler struct
+ *
+ * Inherit from this struct for custom USB functionality
+ */
+struct usbus_handler {
+    struct usbus_handler *next;           /**< List of handlers (to be used by
+                                               @ref usbus_t) */
+    const usbus_handler_driver_t *driver; /**< driver for this handler */
+    usbus_interface_t *iface;             /**< Interface this handler belongs
+                                               to */
+    uint32_t flags;                       /**< Report flags */
+};
+
+/**
+ * @brief helper struct to divide control messages in multiple parts
+ */
+typedef struct {
+    size_t start;                   /**< Start offset of the current part */
+    size_t cur;                     /**< Current position in the message  */
+    size_t len;                     /**< Length of the full message       */
+    size_t transfered;              /**< Number of bytes transfered       */
+    size_t reqlen;                  /**< Maximum length of the request    */
+} usbus_controlslicer_t;
```

Comment:
Should be moved to it's own header file with the function implementations in `usbus_controlslicer.c`

Reply:
Shuffled things around a bit